src/Kestrel.Core/Internal/Infrastructure/TimeoutControl.cs

```diff
@@ -0,0 +1,247 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
+{
+    public class TimeoutControl : ITimeoutControl, IConnectionTimeoutFeature
```

Comment:
Glad to see this decoupled from HttpConnection

Reply:
Spend a few minutes to see what it would take to split this into three types: read timeout, write timeout, and generic timeout. The methods already appear to be distinct and grouped, the biggest overheard looks like it would be passing three types around.

Reply:
I dunno if we should do that. Who does it help? This class is only 200 LOC.

Reply:
It's more of an organizational point, this class represents three district sets of functionality and the caller will only ever be using one at at time.